File: sdk/typescript/src/namespaces/quotas.ts
/**
 * Quotas Namespace API
 *
 * Provides endpoints for managing resource quotas including
 * usage limits, quota policies, and usage tracking.
 */

import type { AragoraClient } from '../client';

/** Quota resource type */
export type QuotaResource = 'debates' | 'agents' | 'storage' | 'api_calls' | 'workflows';

/** Quota definition */
export interface Quota {
  id: string;
  resource: QuotaResource;
  limit: number;
  used: number;
  remaining: number;
  period: 'hourly' | 'daily' | 'monthly';
  resets_at: string;
}

/** Quota policy */
export interface QuotaPolicy {
  id: string;
  name: string;
  description?: string;
  limits: Record<string, number>;
  applies_to: string;
  created_at: string;
}

/** Quota usage history entry */
export interface QuotaUsageEntry {
  timestamp: string;
  resource: string;
  used: number;
  limit: number;
}

/**
 * Quotas namespace for resource limit management.
 *
 * @example
 * ```typescript
 * const quotas = await client.quotas.list();
 * const debateQuota = quotas.find(q => q.resource === 'debates');
 * console.log(`${debateQuota.remaining} debates remaining`);
 * ```
 */
export class QuotasNamespace {
  constructor(private client: AragoraClient) {}

  /** List all quotas for the current workspace. */
  async list(): Promise<Quota[]> {
    const response = await this.client.request<{ quotas: Quota[] }>(
      'GET',
      '/api/v1/quotas'
    );
    return response.quotas;
  }

  /** Get a specific quota by resource. */
  async get(resource: string): Promise<Quota> {
    return this.client.request<Quota>(
      'GET',
      `/api/v1/quotas/${encodeURIComponent(resource)}`
    );
  }

  /** Get quota usage history. */
  async getUsageHistory(options?: {
    resource?: string;
    period?: string;
    limit?: number;
  }): Promise<QuotaUsageEntry[]> {
    const response = await this.client.request<{ usage: QuotaUsageEntry[] }>(
      'GET',
      '/api/v1/quotas/usage',
      { params: options }
    );
    return response.usage;
  }

  /** Update a quota limit (admin only). */
  async updateLimit(resource: string, limit: number): Promise<Quota> {
    return this.client.request<Quota>(
      'PUT',
      `/api/v1/quotas/${encodeURIComponent(resource)}`,
      { body: { limit } }
    );
  }
}

File: sdk/python/aragora_sdk/namespaces/quotas.py
"""
Quotas Namespace API

Provides methods for quota management:
- View current usage and limits
- Get quota details by resource type
- Request quota increases
- Monitor quota consumption history
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


class QuotasAPI:
    """
    Synchronous Quotas API.

    Provides methods for viewing and managing resource quotas including
    API rate limits, storage quotas, and usage limits.

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai")
        >>> quotas = client.quotas.list()
        >>> print(f"Debates used: {quotas['debates']['used']}/{quotas['debates']['limit']}")
        >>> details = client.quotas.get("debates")
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    def list(self) -> dict[str, Any]:
        """
        List all quotas and current usage.

        Returns:
            Dict with quota information for each resource type including
            current usage, limits, and percentage consumed.
        """
        return self._client.request("GET", "/api/v1/quotas")

    def get(self, resource: str) -> dict[str, Any]:
        """
        Get quota details for a specific resource type.

        Args:
            resource: Resource type (e.g., 'debates', 'agents', 'storage',
                'api_calls').

        Returns:
            Dict with detailed quota information including usage history,
            limit, and reset schedule.
        """
        return self._client.request("GET", f"/api/v1/quotas/{resource}")

    def get_usage(self, period: str | None = None) -> dict[str, Any]:
        """
        Get quota usage summary.

        Args:
            period: Time period for usage data (e.g., '1h', '24h', '7d', '30d').

        Returns:
            Dict with usage data across all resource types for the
            specified period.
        """
        params: dict[str, Any] = {}
        if period:
            params["period"] = period
        return self._client.request("GET", "/api/v1/quotas/usage", params=params or None)

    def request_increase(self, resource: str, **kwargs: Any) -> dict[str, Any]:
        """
        Request a quota increase for a resource type.

        Args:
            resource: Resource type to increase quota for.
            **kwargs: Request details including:
                - requested_limit: Desired new limit
                - justification: Reason for the increase

        Returns:
            Dict with request status and ticket ID.
        """
        data: dict[str, Any] = {"resource": resource, **kwargs}
        return self._client.request("POST", "/api/v1/quotas/request-increase", json=data)


class AsyncQuotasAPI:
    """
    Asynchronous Quotas API.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     quotas = await client.quotas.list()
        ...     details = await client.quotas.get("debates")
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    async def list(self) -> dict[str, Any]:
        """List all quotas and current usage."""
        return await self._client.request("GET", "/api/v1/quotas")

    async def get(self, resource: str) -> dict[str, Any]:
        """Get quota details for a specific resource type."""
        return await self._client.request("GET", f"/api/v1/quotas/{resource}")

    async def get_usage(self, period: str | None = None) -> dict[str, Any]:
        """Get quota usage summary."""
        params: dict[str, Any] = {}
        if period:
            params["period"] = period
        return await self._client.request("GET", "/api/v1/quotas/usage", params=params or None)

    async def request_increase(self, resource: str, **kwargs: Any) -> dict[str, Any]:
        """Request a quota increase for a resource type."""
        data: dict[str, Any] = {"resource": resource, **kwargs}
        return await self._client.request("POST", "/api/v1/quotas/request-increase", json=data)
